File: src/generic/sift_keypoint.cpp
#include "searchwrapper.h"
#include <pcl/keypoints/sift_keypoint.h>

#define SIFTKeypoint_ptr pcl::SIFTKeypoint<_PointT, pcl::PointXYZ>::Ptr
#define PointCloud_ptr pcl::PointCloud<_PointT>::Ptr
#define OutputPointCloud_ptr pcl::PointCloud<pcl::PointXYZ>::Ptr

#define KdTree_ptr pcl::KdTreeFLANN<_PointT>::Ptr
#define Octree_ptr pcl::octree::OctreePointCloudSearch<_PointT>::Ptr

PCLIMP(void*, SIFTKeypoint, new)()
{
  return new SIFTKeypoint_ptr(new pcl::SIFTKeypoint<_PointT, pcl::PointXYZ>());
}

PCLIMP(void, SIFTKeypoint, delete)(SIFTKeypoint_ptr *self)
{
  delete self;
}

PCLIMP(void, SIFTKeypoint, setInputCloud)(SIFTKeypoint_ptr *self, PointCloud_ptr *cloud)
{
  (*self)->setInputCloud(*cloud);
}

PCLIMP(void, SIFTKeypoint, setSearchMethod_Octree)(SIFTKeypoint_ptr *self, Octree_ptr *octree)
{
  boost::shared_ptr<OctreeSearchWrapper<_PointT> > wrapper(new OctreeSearchWrapper<_PointT>(*octree));
  (*self)->setSearchMethod(wrapper);
}

PCLIMP(void, SIFTKeypoint, setSearchMethod_KdTree)(SIFTKeypoint_ptr *self, KdTree_ptr *kdtree)
{
  boost::shared_ptr<KdTreeSearchWrapper<_PointT> > wrapper(new KdTreeSearchWrapper<_PointT>(*kdtree));
  (*self)->setSearchMethod(wrapper);
}

PCLIMP(void, SIFTKeypoint, setScales)(SIFTKeypoint_ptr *self, float min_scale, int nr_octaves, int nr_scales_per_octave)
{
  (*self)->setScales(min_scale, nr_octaves, nr_scales_per_octave);
}

PCLIMP(void, SIFTKeypoint, setMinimumContrast)(SIFTKeypoint_ptr *self, float min_contrast)
{
   (*self)->setMinimumContrast(min_contrast);
}

PCLIMP(void, SIFTKeypoint, compute)(SIFTKeypoint_ptr *self, OutputPointCloud_ptr *output)
{
  (*self)->compute(**output);
}

#undef SIFTKeypoint_ptr
#undef PointCloud_ptr
#undef OutputPointCloud_ptr
#undef KdTree_ptr
#undef Octree_ptr

File: SIFTKeypoint.lua
local ffi = require 'ffi'
local torch = require 'torch'
local utils = require 'pcl.utils'
local pcl = require 'pcl.PointTypes'

local SIFTKeypoint = torch.class('pcl.SIFTKeypoint', pcl)

local func_by_type = {}

local function init()

  local SIFTKeypoint_method_names = {
    'new',
    'delete',
    'setInputCloud',
    'setSearchMethod_Octree',
    'setSearchMethod_KdTree',
    'setScales',
    'setMinimumContrast',
    'compute'
  }
  
  for k,v in pairs(utils.type_key_map) do
    func_by_type[k] = utils.create_typed_methods("pcl_SIFTKeypoint_TYPE_KEY_", SIFTKeypoint_method_names, v)
  end   
end

init()

function SIFTKeypoint:__init(pointType)
  pointType = pcl.pointType(pointType or pcl.PointXYZ)
  rawset(self, 'f', func_by_type[pointType])
  self.pointType = pointType
  self.o = self.f.new()
  self:setScales(0.01, 3, 4)
  self:setMinimumContrast(0.001)
end

function SIFTKeypoint:cdata()
  return self.o
end

function SIFTKeypoint:setInputCloud(cloud)
  self.f.setInputCloud(self.o, cloud:cdata())
end

function SIFTKeypoint:setIndices(indices)
  self.f.setIndices(self.o, indices:cdata())
end

function SIFTKeypoint:setSearchMethod(search)
  if torch.isTypeOf(search, pcl.KdTree) then
    self.f.setSearchMethod_KdTree(self.o, search:cdata())
  elseif torch.isTypeOf(search, pcl.Octree) then
    self.f.setSearchMethod_Octree(self.o, search:cdata())
  else
    error("unsupported search method")
  end
end

function SIFTKeypoint:setScales(min_scale, octaves, scales_per_octave)
  self.f.setScales(self.o, min_scale, octaves, scales_per_octave)
end

function SIFTKeypoint:setMinimumContrast(min_contrast)
  self.f.setMinimumContrast(self.o, min_contrast)
end

function SIFTKeypoint:compute(output)
  if not output then
    output = pcl.PointCloud(pcl.Normal)
  end
  self.f.compute(self.o, output:cdata())
  return output
end
